[FEATURE] Default settings templates

burtenshaw opened the pull request:
# Description
<!-- Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

This PR brings back templates to the v2 SDK with a minimal selection of document_classification, response_ranking, and response_rating. When combined with https://github.com/argilla-io/argilla/pull/5418 , this PR will make it easy to load most datasets form the hub in the SDK.

Settings templates can be accessed like this: 

```python
settings = Settings.for_classification(labels=mock_labels)
settings = Settings.for_ranking()
settings = Settings.for_rating()
```

## Options

Settings templates can be used as a base that is added to.

```python
settings = Settings.for_ranking()
settings.fields.add(rg.TextField(name="prompt"))
settings.mapping["x"] = "prompt"
```




We could also 

**Type of change**
<!--  Please delete options that are not relevant. Remember to title the PR according to the type of change  -->
- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**
<!--  Please add some reference about how your feature has been tested.  -->

- Tests are updated

burtenshaw committed:
test: add settings to from hub tests

burtenshaw committed:
feat: implment from hub with settings

burtenshaw committed:
feat: add warning for misaligned datasets

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
Update argilla/tests/integration/test_export_dataset.py

Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
test: expand tests for incorrect settings

burtenshaw committed:
fix: always raise settings error

burtenshaw committed:
fix: import settings for typing

burtenshaw committed:
feat: refactor from hub method to simplify method

burtenshaw committed:
fix: remove image references in tests

burtenshaw committed:
feat: add basic settings templates

burtenshaw committed:
feat: include templates mixin in settings resource

burtenshaw committed:
test: unit tests for settings templates

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

davidberenstein1957 commented on argilla/tests/unit/test_settings/test_settings_templates.py:
```diff
@@ -0,0 +1,35 @@
+
+
+class TestDefaultSettingsMixin:
+    def test_for_document_classification(self):
```
maybe the phrasing `document_classification` is a bit weird? Also, I would also consider adding image cat and token cat as basic defaults.

burtenshaw replied:
Yeah, I can see that the naming is odd. We should iterate on those

My aim is to make a few generic enough methods that could be updated into variations, and avoid an exhaustive list of ML tasks. I.e. `document_classification` could have have alternative fields for image or chat. 

In a subsequent PR, I would like to add these:
- `mapping` in settings. This would make the default templates flexible to column naming. 
- `Features` >> `Settings` : we select and update a 'Settings` class based on the values and names of `datasets.Features`

With these features, the settings templates could be limited because we can get information from the dataset.

davidberenstein1957 replied:
@burtenshaw , perhaps `content_classification` or simply `classification` would be better? Similarly, should we rename to `ranking` / `rating`?

davidberenstein1957 replied:
I agree that we would need a mapping to truly avoid the difficulties because otherwise, any mis-named columns/field/question should still be re-defined, right?

davidberenstein1957 replied:
@burtenshaw, I agree the features to setting would be very interesting.

burtenshaw replied:
> @burtenshaw , perhaps `content_classification` or simply `classification` 

I think these overlap with `token_classification`

> would be better? Similarly, should we rename to `ranking` / `rating`?

This makes sense. What about `for_ranking`,`for_rating`, `for_labels`, `for_spans` ?

davidberenstein1957 reviewed (commented):
I loved the concept myself but I am not 100% sure where this would fit in the docs, and if these config options still are an issue with the renewed singular docs.

burtenshaw committed:
fix: check if id exists when logging

burtenshaw committed:
feat: use repo_id as name

burtenshaw committed:
test: update tests to use random names

burtenshaw committed:
respond: improve imports and warnings

burtenshaw committed:
[FEATURE] [WIP] Persist mapping in Settings (#5428)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR builds on https://github.com/argilla-io/argilla/pull/5426

It expands `rg.Settings` to take a mapping parameter which is passed to
`dataset.records`. The purpose of this is
- to store mappings with settings so that records can be ingested using
the mapping after persistence.
- to make it easier to share argilla datasets on the hub
- to make it easier to implement templates for common use cases.

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

burtenshaw committed:
Merge branch 'refactor/simplify-casting-in-from_hub' into feature/settings-templates

burtenshaw committed:
Merge branch 'feature/settings-templates' of https://github.com/argilla-io/argilla into feature/settings-templates

burtenshaw committed:
test: fix tests for conflicting datasets

burtenshaw committed:
feat: add generic mappings to templates

burtenshaw committed:
test: fix tests to take labels parameter

burtenshaw committed:
Merge branch 'develop' into feature/settings-templates

frascuchon commented on argilla/src/argilla/settings/_templates.py:
```diff
@@ -0,0 +1,159 @@
+            guidelines="Select a label for the document.",
+            fields=[TextField(name="text")],
+            questions=[LabelQuestion(name="label", labels=labels)],
+            mapping={"input": "text", "output": "label", "document": "text"},
```
What about parameterize names on template creation? Diving in the code is a bit hard to follow all this mappings at the settings level. 

````python
Settings.for_document_classification(
   fields=["text", "title",...],
....
)
```

frascuchon replied:
Also, maybe this can be more flexible for other dataset structure.

frascuchon commented on argilla/src/argilla/settings/_templates.py:
```diff
@@ -0,0 +1,159 @@
+            fields=[TextField(name="instruction"), TextField(name="response1"), TextField(name="response2")],
+            questions=[RankingQuestion(name="ranking", values=["response1", "response2"])],
+            mapping={
+                "input": "instruction",
+                "prompt": "instruction",
```
Same comment here. Maybe users want to see both fields separately. 

```python
Settings.for_response_ranking(
   instruction=["input", "prompt"], 
   response1="chosen", 
   response2="rejected"
)
```

frascuchon commented on argilla/src/argilla/settings/_templates.py:
```diff
@@ -0,0 +1,159 @@
+    def for_response_rating(
+        cls: "Settings",
+        guidelines: Optional[str] = None,
+        fields: Optional[List["Field"]] = None,
```
What's the purpose of these `fields` argument? I can imagine some kind of template extension, but maybe easier if we pass just the names? Users can always change them after the settings creation

burtenshaw committed:
feat: simplify interface and support different field types

burtenshaw committed:
test: update tests

burtenshaw committed:
Merge branch 'develop' into feature/settings-templates

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
Merge branch 'develop' into feature/settings-templates

frascuchon commented on argilla/src/argilla/settings/_templates.py:
```diff
@@ -0,0 +1,130 @@
+    from argilla.settings._field import TextField
+
+
+def _get_field_type(field_type: str) -> "TextField":
```
```suggestion
def _get_field_by_type(field_type: str) -> "TextField":
```

frascuchon commented on argilla/src/argilla/settings/_templates.py:
```diff
@@ -0,0 +1,130 @@
+    @classmethod
+    def for_ranking(
+        cls: "Settings",
+        field_type: Optional[Literal["text", "image", "chat"]] = "text",
```
"chat" value shouldn't be here yet, right?

frascuchon commented on argilla/src/argilla/settings/_templates.py:
```diff
@@ -0,0 +1,130 @@
+    @classmethod
+    def for_rating(
+        cls: "Settings",
+        field_type: Optional[Literal["text", "image", "chat"]] = "text",
```
ditto
```suggestion
        field_type: Optional[Literal["text", "image"]] = "text",
```

burtenshaw committed:
feat: remove chat as an optional field type

Co-authored-by: Paco Aranda <frascuchon@gmail.com>

burtenshaw committed:
feat: improve naming
